include/rmm/mr/device/detail/arena.hpp

```diff
@@ -0,0 +1,616 @@
+  bool is_valid() const { return pointer != nullptr; }
+
+  /// Returns true if this block is a superblock, false otherwise.
+  bool is_superblock() const { return is_head && size == superblock_size; }
+
+  /**
+   * @brief Verifies whether this block can be merged to the beginning of block b.
+   *
+   * @param b The block to check for contiguity.
+   * @return true Returns true if this block's `pointer` + `size` == `b.ptr`, and `not b.is_head`,
+                  false otherwise.
+   */
+  bool is_contiguous_before(block const& b) const
+  {
+    return increment(pointer, size) == b.pointer and not(b.is_head);
+  }
+
+  /**
+   * @brief Is this block large enough to fit `sz` bytes?
+   *
+   * @param sz The size in bytes to check for fit.
+   * @return true if this block is at least `sz` bytes.
+   */
+  bool fits(std::size_t sz) const { return size >= sz; }
+
+  /**
+   * @brief Split this block into two by the given size.
+   *
+   * @param sz The size in bytes of the first block.
+   * @return std::pair<block, block> A pair of blocks split by sz.
+   */
+  std::pair<block, block> split(std::size_t sz) const
+  {
+    RMM_LOGGING_ASSERT(size >= sz);
+    if (size > sz) {
+      return {{pointer, sz, is_head}, {increment(pointer, sz), size - sz, false}};
+    } else {
+      return {*this, {}};
+    }
+  }
+
+  /**
+   * @brief Coalesce two contiguous blocks into one.
+   *
+   * `this` must immediately precede `b` and both `this` and `b` must be from the same superblock.
+   *
+   * @param b block to merge.
+   * @return block The merged block.
+   */
+  block merge(block const& b) const
+  {
+    RMM_LOGGING_ASSERT(is_contiguous_before(b));
+    return {pointer, size + b.size, is_head};
+  }
+
+  /**
+   * @brief Increment the given pointer by the given size.
+   *
+   * @param pointer the pointer to increment.
+   * @param size the size to add to the pointer.
+   * @return void* The resulting pointer.
+   */
+  static void* increment(void* pointer, std::size_t size)
+  {
+    return static_cast<void*>(static_cast<char*>(pointer) + size);
+  }
+
+  /// Used by std::set to compare blocks.
+  bool operator<(block const& b) const { return pointer < b.pointer; }
+};
+
+/// The required allocation alignment.
+constexpr std::size_t allocation_alignment = 256;
+
+/**
+ * @brief Align up to the allocation alignment.
+ *
+ * @param[in] v value to align
+ * @return Return the aligned value
+ */
+constexpr std::size_t align_up(std::size_t v) noexcept
+{
+  return rmm::detail::align_up(v, allocation_alignment);
+}
+
+/**
+ * @brief Align down to the allocation alignment.
+ *
+ * @param[in] v value to align
+ * @return Return the aligned value
+ */
+constexpr std::size_t align_down(std::size_t v) noexcept
+{
+  return rmm::detail::align_down(v, allocation_alignment);
+}
+
+/**
+ * @brief Get the first free block of at least `size` bytes.
+ *
+ * Address-ordered first-fit has shown to perform slightly better than best-fit when it comes to
```

Comment:
It is not intuitive that first-fit would reduce memory fragmentation compared to best-fit. Best fit finds the tightest fitting block, after all.  Do you have measurements of fragmentation?

Reply:
This paper has a bunch of experiments (https://www.cs.tufts.edu/~nr/cs257/archive/paul-wilson/fragmentation.pdf). It's pretty old, but some of the test programs (e.g. gcc) are still in use. As I said addressed-ordered first fit is only slightly better than best fit for fragmentation, but it's cheaper to run since we don't have to always scan the full free list.

They didn't fully understand why, but provided some theories:
>The best-fit policy tries to preferentially use small free blocks over large free blocks. This characteristic gives the neighbors of the large free blocks more time to die and be merged into yet larger free blocks, which, in turn, makes them even less likely that best fit will allocate something out of these larger free blocks. The cycle continues until there are only a few very large areas of contiguous free memory out of which to allocate free blocks. When one of these free blocks is used for memory allocation, a small piece is split out of it, making it somewhat smaller, which will make it more likely that that same free block will be used for subsequent memory requests, saving the other larger free areas for later needs. 
>
>Using address-ordered free lists, which worked so well for first fit and next fit, can be viewed as a variation on this same theme. Blocks at one end of memory are used preferentially over blocks at the other end. This gives objects at the end of memory from which new blocks are not being allocated more time to die and merge with their neighbors. Note, however, that this theme is much stronger with first fit address ordered than with next fit address ordered. We believe this is why first fit address ordered performs much better than next fit address ordered.

Reply:
Thanks. I saw that paper and skimmed it but I thought that it was showing lower fragmentation for best-fit. Here's a better-quality PDF link: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.97.5185&rep=rep1&type=pdf